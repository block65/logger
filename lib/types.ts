import { AsyncLocalStorage } from 'async_hooks';
import type pino from 'pino';
import type { MixinFn, MixinFnWithData } from './mixins.js';
import type { PrettyTransportOptions } from './pretty-transport.js';
import type { SentryTransportOptions } from './sentry-transport.js';

export type Falsy = false | undefined | null;

export enum LogLevelNumbers {
  Fatal = 60,
  Error = 50,
  Warn = 40,
  Info = 30,
  Debug = 20,
  Trace = 10,
  Silent = Infinity,
}

export type LevelWithSilent = pino.LevelWithSilent;

export interface LogDescriptor {
  level: LogLevelNumbers;
  time: number;

  msg?: string;
  name?: string | undefined;

  // maye not be available depending on platform/cli
  pid?: number;
  hostname?: string;

  [key: string]: unknown;
}

type RemoveIndex<T> = {
  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
};

// pino tacks on a nasty `Record<string, any>` to its logger type
// we try to remove it here
export type BaseLogger = RemoveIndex<pino.Logger>;

export interface Logger extends BaseLogger {
  als: AsyncLocalStorage<AlsContext>;
  flushTransports: () => Promise<void>;
}

export interface AlsContext {
  id: string;
  context?: Record<string, unknown>;
}

export interface AlsContextOutput {
  _contextId: AlsContext['id'];
  _context?: AlsContext['context'];
}

export type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions
  extends Omit<
    pino.LoggerOptions,
    'mixin' | 'prettyPrint' | 'prettifier' | 'color' | 'transport' | 'level'
  > {
  prettyOptions?: boolean | PrettyTransportOptions;
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | MixinFn | Falsy)[];
  transports?: pino.TransportMultiOptions['targets'];
  sentryTransportOptions?: SentryTransportOptions;
  level?: pino.LevelWithSilent;
}
