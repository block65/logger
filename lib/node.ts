import type { RequestHandler } from 'express';
import { serializeError } from 'serialize-error';
import * as pino from 'pino';
import * as pinoHttp from 'pino-http';
import { createNamespace, Namespace } from 'cls-hooked';
import { hostname } from 'os';
import { randomBytes } from 'crypto';
import { IncomingMessage, ServerResponse } from 'http';
import { sep } from 'path';

type Falsy = false | undefined | null;
type MixinFnWithData = (
  data: ReturnType<pino.MixinFn>,
) => ReturnType<pino.MixinFn>;

export type Logger = {
  cls: Namespace;
} & pino.Logger;

interface ClsContext {
  _contextId: string;
  _context?: Record<string, unknown>;
}

type NamespaceContext = {
  id?: unknown;
  _ns_name?: unknown;
} & Partial<ClsContext>;

type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions extends Omit<pino.LoggerOptions, 'mixin'> {
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | pino.MixinFn | Falsy)[];
}

const suggestedHostname = hostname();

const defaultLoggerOptions: pino.LoggerOptions = {
  level: 'info',
  serializers: {
    err: serializeError,
  },
  base: {
    pid: process.pid,
    // in the case we get localhost (gcp cloud run), use random bytes instead
    // this helps us debug if we need to isolate calls running in a specific
    // container or instance (one which stays alive between invocations)
    hostname:
      suggestedHostname !== 'localhost'
        ? suggestedHostname
        : `zz${randomBytes(5).toString('hex')}`, // zz indicates synthetic hostname
  },
};

const gcpLevelMap: Record<string, string> = {
  trace: 'DEFAULT',
  debug: 'DEBUG',
  info: 'INFO',
  warn: 'WARNING',
  error: 'ERROR',
  fatal: 'CRITICAL',
};

function getPlatformLoggerOptions(
  platform?: ComputePlatform,
): pino.LoggerOptions {
  switch (platform) {
    case 'gcp-cloudrun':
      return {
        level: 'info',
        // customLevels: {
        //   DEBUG: 100,
        //   INFO: 200,
        //   NOTICE: 300,
        //   WARNING: 400,
        //   ERROR: 500,
        //   CRITICAL: 600,
        //   ALERT: 700,
        //   EMERGENCY: 800,
        // },
        messageKey: 'message',
        formatters: {
          level(levelLabel /* , levelNumber */) {
            return { severity: gcpLevelMap[levelLabel] };
          },
        },
        timestamp: pino.stdTimeFunctions.isoTime,
      };
    case 'aws':
    case 'aws-lambda':
    default:
      return {};
  }
}

let counter = 0;

const isDevelopment = process.env.NODE_ENV === 'development';

function callerMixin(): { caller: string | undefined } {
  return {
    caller: Error()
      .stack?.split('\n')
      .slice(2)
      .filter(
        (s) =>
          // !s.includes(`api${sep}src${sep}app`) &&
          // !s.includes(
          //   `node_modules${sep}express${sep}lib${sep}router`,
          // ) &&
          !s.includes(`node_modules${sep}pino`) &&
          !s.includes(`block65${sep}logger`),
      )?.[0]
      ?.substr(7),
  };
}

function composeMixins(
  mixins: (MixinFnWithData | pino.MixinFn | Falsy)[],
): pino.MixinFn {
  return () =>
    mixins.reduce((accum, mixin) => {
      if (!mixin) {
        return accum;
      }
      return {
        ...accum,
        ...mixin(accum),
      };
    }, {});
}

function createContextMixin(cls: Namespace): pino.MixinFn {
  return (): Partial<ClsContext> => {
    const { active }: { active: NamespaceContext | null } = cls || {};

    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { id, _ns_name, ...clsContext } = active || {};
    return clsContext;
  };
}

export function createLogger(
  opts: CreateLoggerOptions = {},
  stream?: pino.DestinationStream,
  // destination?: pino.DestinationStream,
): Logger {
  const cls = createNamespace(`logger/${counter}`);
  counter += 1;

  const {
    platform,
    traceCaller = isDevelopment,
    mixins = [],
    ...userPinoOpts
  } = opts;

  const platformLoggerOptions = getPlatformLoggerOptions(platform);

  const mixin = composeMixins([
    ...mixins,
    createContextMixin(cls),
    traceCaller && callerMixin,
  ]);

  const pinoInstance = stream
    ? pino(
        {
          ...defaultLoggerOptions,
          ...platformLoggerOptions,
          ...userPinoOpts,
          mixin,
        },
        stream,
      )
    : pino({
        ...defaultLoggerOptions,
        ...platformLoggerOptions,
        ...userPinoOpts,
        mixin,
      });

  return Object.create(pinoInstance, {
    cls: {
      value: cls,
      configurable: false,
    },
  });
}

export function expressLogger(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): RequestHandler {
  return (req, res, next): void => {
    res.set('x-context-id', contextId);

    namespace.run(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      next();
    });
  };
}

export function lambdaLogger<T extends (...args: any[]) => any>(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): (next: T) => (fn: T) => ReturnType<T> {
  return (fn) => {
    return namespace.runAndReturn(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      return fn();
    });
  };
}

export function createHttpLogger(
  logger: pino.Logger,
  opts: Omit<pinoHttp.Options, 'logger' | 'customLogLevel'> = {},
): pinoHttp.HttpLogger {
  return pinoHttp({
    ...opts,
    genReqId: () => ({}),
    logger,
    customSuccessMessage(res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customErrorMessage(err: Error, res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customLogLevel(res: ServerResponse, err: Error) {
      if (res.statusCode >= 500 || err) {
        return 'error';
      }
      if (res.statusCode >= 400 && res.statusCode < 500) {
        return 'warn';
      }
      return 'trace';
    },
    reqCustomProps(req: IncomingMessage) {
      return { url: req.url };
    },
  });
}
