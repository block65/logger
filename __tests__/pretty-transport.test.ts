import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { randomBytes } from 'node:crypto';
import { FileHandle, open, readFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { PassThrough } from 'node:stream';
import { pipeline } from 'node:stream/promises';
import { TextDecoder } from 'node:util';
import { LogLevelNumbers } from '../lib/types.js';
import { createTmpLogfileDest, writeLogsToStream } from './helpers.js';

describe('Pretty Transport', () => {
  const oldEnv = process.env;

  beforeEach(() => {
    jest.resetModules();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
    process.env = { ...oldEnv };
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('pretty logger no-color', async () => {
    process.env.NO_COLOR = 'true';
    const { prettyTransport } = await import('../lib/pretty-transport.js');

    const [destination, getLogs] = await createTmpLogfileDest();
    const transport = await prettyTransport({ destination });

    await writeLogsToStream(transport, {
      level: LogLevelNumbers.Info,
      time: Date.now(),
      pid: 1,
      hostname: 'test',
      msg: 'kek',
      random: {
        bool: true,
        str: 'hello',
        number: 12,
        arr: [1, 2, 3],
      },
    });

    const logs = await getLogs();

    expect(logs).toMatchSnapshot();
  });

  test('pretty logger force-color', async () => {
    const { prettyTransport } = await import('../lib/pretty-transport.js');

    const [destination, getLogs] = await createTmpLogfileDest();

    const transport = await prettyTransport({
      destination,
      color: true,
    });

    await writeLogsToStream(transport, {
      level: LogLevelNumbers.Error,
      err: {
        message: 'hello',
      },
      time: Date.now(),
    });
    const logs = await getLogs();

    expect(logs).toMatchSnapshot();
  });
});
