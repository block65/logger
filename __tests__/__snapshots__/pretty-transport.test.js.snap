// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty Transport pretty logger force-color 1`] = `
"[90m2009-02-13T23:31:30.000Z[39m [31mERROR[39m  { err: { message: [32m'hello'[39m } }
"
`;

exports[`Pretty Transport pretty logger no-color 1`] = `
"2009-02-13T23:31:30.000Z INFO    kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
"
`;
